modules/utxo-lib/src/bitgo/psbt/fromHalfSigned.ts

```diff
@@ -1,145 +1,88 @@
-      const publicKeys = payments.p2tr_ns({ output: script }, { eccLib }).pubkeys;
-      if (!publicKeys || publicKeys.length !== 2) {
-        throw new Error('expected 2 pubkeys');
-      }
-      if (witness.length !== 2) {
-        throw new Error(`expected exactly 2 signatures, got ${witness.length}`);
-      }
-      update.tapScriptSig = [];
-      let signature;
-      while ((signature = witness.pop()) !== undefined) {
-        if (signature.length === 0) {
-          publicKeys.shift(); // No signature for this key
-          continue;
-        }
-        const leafHash = taproot.getTapleafHash(eccLib, controlBlock, script);
-        const pubkey = publicKeys.shift();
-        if (!pubkey) throw new Error("Impossible, known 2-length things didn't match");
-        update.tapScriptSig.push({ signature, pubkey, leafHash });
-      }
-    } else if (isP2wsh(prevOut.script, redeemScript)) {
-      const witnessScript = witness.pop();
-      if (!witnessScript) throw new Error('Invalid witness structure');
-      update.witnessScript = witnessScript;
-      const decompiledWitnessScript = bscript.decompile(witnessScript);
-      if (!decompiledWitnessScript) {
-        throw new Error('Invalid witnessScript, failed to decompile');
-      }
-      const hashFn = (hashType) => tx.hashForWitnessV0(vin, witnessScript, prevOut.value, hashType);
-      update.partialSig = matchSignatures(hashFn, witnessScript, witness);
-    } else if (isP2wpkh(prevOut.script, redeemScript)) {
-      if (witness.length === 2) {
-        const pubkey = witness.pop();
-        const signature = witness.pop();
-        if (!Buffer.isBuffer(pubkey) || !Buffer.isBuffer(signature)) {
-          throw new Error('Invalid pubkey or signature');
-        }
-        update.partialSig = [{ pubkey, signature }];
+
+  if (!hasWitnessData(parsed.scriptType) && !nonWitnessUtxo) {
+    throw new Error(`scriptType ${parsed.scriptType} requires prevTx Buffer`);
+  }
+
+  switch (parsed.scriptType) {
+    case 'p2shP2pk':
+      return {
+        nonWitnessUtxo,
+        partialSig: [{ pubkey: parsed.publicKeys[0], signature: parsed.signatures[0] }],
+      };
+    case 'p2sh':
+    case 'p2wsh':
+    case 'p2shP2wsh':
+      return omitUndefined({
+        nonWitnessUtxo,
+        partialSig: getPartialSigs(),
+        redeemScript: parsed.redeemScript,
+        witnessScript: parsed.witnessScript,
+      });
+    case 'p2tr':
+      if (!('controlBlock' in parsed)) {
+        throw new Error(`keypath not implemented`);
       }
-      return update;
-    }
+      return {
+        tapLeafScript: [
+          {
+            controlBlock: parsed.controlBlock,
+            script: parsed.pubScript,
+            leafVersion: parsed.leafVersion,
+          },
+        ],
+        tapScriptSig: getPartialSigs().map((obj) => ({
+          ...obj,
+          leafHash: taproot.getTapleafHash(eccLib, parsed.controlBlock, parsed.pubScript),
```

Comment:
maybe define `leafHash` before the return statement, rather than re-hashing each loop? I know it's only 1 or 2, but...